File: 08/main.js
#!/usr/bin/env node
const regex = /^([a-z]{3}) ([-+]\d+)$/;
const input = require('fs')
  .readFileSync(
    process.argv[2] || require('path').join(__dirname, 'input.txt'),
    'utf8'
  )
  .trim()
  .split('\n')
  .map(line => {
    const match = regex.exec(line);
    if (match === null) {
      throw new Error(`Invalid line: '${line}'`);
    }
    return {
      type: match[1],
      value: +match[2],
      executed: false,
    };
  });


const run = (instructions, reset = true) => {
  let accumulator = 0;
  let index = 0;
  let prevIndex = -1;
  let infinite = false;
  if (reset) {
    for (const ins of instructions) {
      ins.executed = false;
    }
  }
  while (true) {
    const ins = instructions[index];
    if (ins === undefined) {
      break;
    }
    if (ins.executed) {
      infinite = true;
      break;
    }
    ins.executed = true;

    prevIndex = index;
    switch (ins.type) {
      case 'acc':
        accumulator += ins.value;
      case 'nop':
        ++index;
        break;
      case 'jmp':
        index += ins.value;
        break;
      default:
        throw new Error(`Invalid instruction type: '${ins.type}'`);
    }
  }

  return [accumulator, infinite];
};


const partTwo = instructions => {
  for (const ins of instructions) {
    let prevType = ins.type;
    switch (ins.type) {
      case 'acc':
        continue;
      case 'jmp':
        ins.type = 'nop';
        break;
      case 'nop':
        ins.type = 'jmp';
        break;
    }
    const [accumulator, infinite] = run(instructions, true);
    // console.log(
    //   'Swapped %d: %s -> %s: %d (%s)',
    //   i, prevType, ins.type, accumulator, infinite);
    if (!infinite) {
      return accumulator;
    }
    ins.type = prevType;
  }
  return null;
};


module.exports.input = input;
// console.log(input);
console.log('Part one:', run(input, false)[0]);
console.log('Part two:', partTwo(input));

File: 08/main.cpp
#include <cstdlib>
#include <iostream>
#include <fstream>
#include <vector>
#include <utility>


struct Instruction {
  enum class Type: unsigned char {
    nop = 0,
    acc,
    jmp
  };

  Instruction(
    int value = 0,
    Type type = Type::nop,
    bool executed = false
  )
    : m_value(value)
    , m_type(type)
    , m_executed(executed)
  {
  }

  inline void flip() noexcept {
    m_type = static_cast<Type>(2 - static_cast<unsigned char>(m_type));
  }

  int m_value;
  Type m_type;
  bool m_executed;
};


template <typename InputStream>
InputStream& operator>>(InputStream& input, Instruction& ins) {
  std::string typeStr;
  input >> typeStr;
  if (!input.good()) {
    throw std::invalid_argument("Invalid input");
  }

  Instruction tmp;
  if (typeStr == "acc") {
    tmp.m_type = Instruction::Type::acc;
  }
  else if (typeStr == "jmp") {
    tmp.m_type = Instruction::Type::jmp;
  }
  else if (typeStr == "nop") {
    tmp.m_type = Instruction::Type::nop;
  }
  else {
    throw std::invalid_argument(typeStr);
  }

  input >> tmp.m_value;
  std::swap(ins, tmp);
  return input;
}


constexpr bool INFINITE_LOOP = true;
constexpr bool PROGRAM_TERMINATES = false;


template <typename RAIter>
std::pair<int, bool>
run(RAIter begin, RAIter end) {
  int accumulator = 0;
  for (auto current = begin; current != end; ) {
    if (std::exchange(current->m_executed, true)) {
      return {accumulator, INFINITE_LOOP};
    }
    switch (current->m_type) {
      case Instruction::Type::nop:
        ++current;
        break;
      case Instruction::Type::acc:
        accumulator += current->m_value;
        ++current;
        break;
      case Instruction::Type::jmp:
        current += current->m_value;
        break;
    }
  }
  return {accumulator, PROGRAM_TERMINATES};
}


int main(int const argc, char const* const* const argv) {
  std::ifstream ifs(argc < 2 ? DEFAULT_INPUT_FILE : argv[1]);
  std::vector<Instruction> instructions;
  while (ifs.good() && ifs.peek() != std::ifstream::traits_type::eof()) {
    Instruction ins;
    ifs >> ins >> std::ws;
    instructions.emplace_back(ins);
  }

  auto const partOne = run(
      std::begin(instructions),
      std::end(instructions));

  int partTwo = -1;
  for (auto& ins: instructions) {
    for (auto& instruction: instructions) {
      instruction.m_executed = false;
    }
    ins.flip();
    auto const [accumulator, infinite] = run(
        std::begin(instructions),
        std::end(instructions));
    if (!infinite) {
      partTwo = accumulator;
    }
    ins.flip();
  }

  std::cout << "Part one: " << partOne.first << std::endl;
  std::cout << "Part two: " << partTwo << std::endl;

  return 0;
}
